oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/Jwt.java

```diff
@@ -80,4 +85,128 @@ public Jwt(String tokenValue, Instant issuedAt, Instant expiresAt,
 	public Map<String, Object> getClaims() {
 		return this.claims;
 	}
+	
+	/**
+	 * Helps configure a {@link Jwt}
+	 *
+	 * @author Jérôme Wacongne &lt;ch4mp&#64;c4-soft.com&gt;
+	 */
+	public static class Builder implements JwtClaimAccessor {
+		private String tokenValue;
+		private final Map<String, Object> claims = new HashMap<>();
+		private final Map<String, Object> headers = new HashMap<>();
+
+		public Builder tokenValue(String tokenValue) {
+			this.tokenValue = tokenValue;
+			return this;
+		}
+
+		public Builder claim(String name, Object value) {
+			this.claims.put(name, value);
+			return this;
+		}
+
+		public Builder clearClaims(Map<String, Object> claims) {
+			this.claims.clear();
+			return this;
+		}
+
+		/**
+		 * Adds to existing claims (does not replace existing ones)
+		 * @param claims claims to add
+		 * @return this builder to further configure
+		 */
+		public Builder claims(Map<String, Object> claims) {
+			this.claims.putAll(claims);
+			return this;
+		}
+
+		public Builder header(String name, Object value) {
+			this.headers.put(name, value);
+			return this;
+		}
+
+		public Builder clearHeaders(Map<String, Object> headers) {
+			this.headers.clear();
+			return this;
+		}
+
+		/**
+		 * Adds to existing headers (does not replace existing ones)
+		 * @param headers headers to add
+		 * @return this builder to further configure
+		 */
+		public Builder headers(Map<String, Object> headers) {
+			headers.entrySet().stream().forEach(e -> this.header(e.getKey(), e.getValue()));
+			return this;
+		}
+
+		public Jwt build() {
+			Assert.isTrue(hasTokenValue(), "token value must be set");
+			Assert.isTrue(hasName(), "name must be set");
+			Assert.isTrue(hasHeader(), "at least one header must be set");
+			return new Jwt(
+					this.tokenValue,
+					getClaimAsInstant(JwtClaimNames.IAT),
+					getClaimAsInstant(JwtClaimNames.EXP),
+					new HashMap<>(this.headers),
```

Comment:
I don't like to rely on an other class implementation for a method to work as expected but as I moved this code in a Jwt inner class, I did the optimisation. I also added a test, just in case someone touches Jwt constructor...